NIP-5C: Scrolls (WASM programs) (#2281)
nostr-protocol/nips: 5C.md

# NIP-5C

## Scrolls

`draft` `optional`

> This NIP defines a standard for publishing self-contained WebAssembly programs as Nostr events ("scrolls"). Scrolls are composed of basic metadata (`name`, `description` and `image`) and initial execution parameters defined as tags and a WASM binary, base64-encoded as the content. The binaries are executed in a sandbox inside a "host" (i.e. a proper Nostr client). Scrolls interact with Nostr only through a set of simple APIs provided by the host.

Reviewer: ```suggestion
This NIP defines a standard for publishing self-contained WebAssembly programs as Nostr events ("scrolls"). Scrolls are composed of basic metadata (`name`, `description` and `icon`) and initial execution parameters defined as tags and a WASM binary, base64-encoded as the content. The binaries are executed in a sandbox inside a "host" (i.e. a proper Nostr client). Scrolls interact with Nostr only through a set of simple APIs provided by the host.
```

      ["param", "me", "myself", "public_key", "required"],
      [
        "param",
        "public_key",
        "person with whom I want to see my interactions",
        "public_key",
        "required"
      ]
    ]
}
```

### Parameter Passing

When the host calls `run()`, it passes a single pointer to a buffer containing all parameters in the order they appear in the tags. Each type has a specific encoding:

| Type         | Encoding                                                              | Size          |
| ------------ | --------------------------------------------------------------------- | ------------- |
| `public_key` | 32 bytes (should all be set to zero if the parameter is not provided) | 32 bytes      |
| `event`      | i32 handle                                                            | 4 bytes       |
| `string`     | u32_be length followed by UTF-8 bytes                                 | 4 + len bytes |
| `number`     | i32                                                                   | 4 bytes       |
| `timestamp`  | unix timestamp as u32_be                                              | 4 bytes       |

> | `relay`      | relay URL, same as string                                             | 4 + len bytes |

Reviewer: If it's not required and not provided how will it looks like?

For `public_key` it's clear, it will be 32 zeroed bytes, what about the other types?

Reviewer: Something like this would looks good.

```suggestion
### Parameter Passing

When the host calls `run()`, it passes a single pointer to a buffer containing all parameters in the order they appear in the tags. Each type has a specific encoding:

| Type         | Encoding                                                              | Size          | If not provided (for not required params ) |
| ------------ | --------------------------------------------------------------------- | ------------- | ------------------------------------------ |
| `public_key` | 32 bytes                                                              | 32 bytes      | 32 zeroed bytes                            |
| `event`      | i32 handle                                                            | 4 bytes       | 4 zeroed bytes                             |
| `string`     | u32_be length followed by UTF-8 bytes                                 | 4 + len bytes | 4 zeroed bytes                             |
| `number`     | i32                                                                   | 4 bytes       | 4 zeroed bytes                             |
| `timestamp`  | unix timestamp as u32_be                                              | 4 bytes       | 4 zeroed bytes                             |
| `relay`      | relay URL, same as string                                             | 4 + len bytes | 4 zeroed bytes                             |
```

Reviewer: For example, this is my Rust code for reading an optional string, and it's easy to work with (using zeroed bytes). The only problem is that it treats the empty string as None, which I think is acceptable, who would ask for an empty string anyway?

```rust
impl<'a, 'b> ReadParam<'a, 'b> for Option<&'b str> {
    fn read_param(cursor: &'a mut usize, buffer: &'b [u8]) -> Self {
        if buffer.len() < *cursor + 4 {
            panic!(
                "ReadParam(Option<str>): Out of bounds reading length prefix at cursor {cursor} (buffer len: {})",
                buffer.len()
            );
        }

        if buffer[*cursor..*cursor + 4] != [0; 4] {
            return Some(<&str as ReadParam>::read_param(cursor, buffer));
        }

        *cursor += 4;
        None
    }
}
```

Reviewer: Oh wait, it will read 0 number as None while it's 0 :) is this okay? should we use `0xFFFFFFFF` (i32::MAX) as None/nil?

I think we should use `0xFFFFFFFF` if the value is not required and not provided

Reviewer: `0xFFFFFFFF` is not good, because it's `-1` and the program can take it as a number.

I'm thinking about adding a bit in the parameter position, it will indicate whatever there is a param you have to read or there is not and go to the next one (if your program required it you panic). `0` means no parameter provided, while `1` means there is a parameter you can read.

So, if your program take the event ID as an optional and a number, you will receive them like this `01<number-data>`, if the event is there but the number is optional, it will be `1<event-data>0`. So, just prefix the value with `1` if it's there or replace it with `0` if not.

Reviewer: this is actually cool, the program can create a function that read a single byte from the buffer and returns `true` if it's `1` or `false` if it's `0` (it will add 1 to the cursor) then the program can read the value based on this byte.

Let me know if you like this.

Reviewer: @fiatjaf the new version would looks like this

---

### Parameter Passing

When the host calls `run()`, it passes a single pointer to a buffer containing all parameters in the order they appear in the tags. Each type has a specific length or a way to determine its length (for strings, as described in "String Convention", this is a 4-byte prefix).

The host **must** prefix every parameter with a single presence byte: `1` indicates the parameter is provided, and `0` indicates it is omitted. If a required parameter is prefixed with `0`, the program must panic.

| Type         | Encoding                                                              | Size (if provided) |
| ------------ | --------------------------------------------------------------------- | ------------------ |
| `public_key` | 32 bytes                                                              | 32 bytes           |
| `event`      | i32 handle                                                            | 4 bytes            |
| `string`     | u32_be length followed by UTF-8 bytes                                 | 4 + len bytes      |
| `number`     | i32                                                                   | 4 bytes            |
| `timestamp`  | Unix timestamp as u32_be                                              | 4 bytes            |
| `relay`      | Relay URL, same encoding as string                                    | 4 + len bytes      |

For example, in a scroll with these parameters:
```json
[
    ["param", "me", "", "public_key", "required"],
    ["param", "target event", "", "event", "required"],
    ["param", "target public key", "", "public_key", ""],
    ["param", "target relay", "", "relay", "required"]
]
```

The buffer layout would be:

```text
1[ 32-byte pubkey ] 1[ 4-byte handle ] 0 1[ 4-byte len ][ UTF-8 relay URL ]
```

Because `target public key` is optional and not provided, its position contains only the `0` presence byte.

### `nostr.req_new() -> i32`

Create a new empty request handle.

### `nostr.req_add_author(req: i32, pubkey_ptr: i32) -> void`

Add a pubkey to the `authors` filter. It must be a pointer to a 32-byte buffer.

### `nostr.req_add_author_hex(req: i32, pubkey_hex_ptr: i32) -> void`

Same as `req_add_author` but with a hex-encoded pubkey string. Length is assumed to be 64 characters.

### `nostr.req_add_id(req: i32, id_ptr: i32) -> void`

Add an id to the `ids` filter. It must be a pointer to a 32-byte buffer.

### `nostr.req_add_id_hex(req: i32, id_hex_ptr: i32) -> void`

Same as `req_add_id` but with a hex-encoded id string. Length is assumed to be 64 characters.

### `nostr.req_add_kind(req: i32, kind: i32) -> void`

Add a kind integer to the `kinds` filter.

### `nostr.req_add_tag(req: i32, tag_ptr: i32, tag_len: i32, value_ptr: i32, value_len: i32) -> void`

> Add a value to a tag filter. `tag_ptr/tag_len` points to the tag name (e.g. `"p"` -- will be added to the filter as `"#p"`); `value_ptr/value_len` points to the tag value to match. The value is treated as a string.

Reviewer: This is a bug both here and in the runtime. According to NIP-01, the tag should be a single-letter string.

```ts
req_add_tag: (
        req: number,
        tPtr: number,
        tLen: number,
        vPtr: number,
        vLen: number
      ): number => {
        const tag = readStr(tPtr, tLen)
        const value = readStr(vPtr, vLen)
        const r = get(req, "req")
        const tags = r.filter[`#${tag}`] || []
        pushUnique(tags, value)
        r.filter[`#${tag}`] = tags
        console.log(
          `[nprogram] req_add_tag(req=${req}, tag="${tag}", value="${value.slice(0, 32)}...")`
        )
        return 0
      },
```

```json
"#<single-letter (a-zA-Z)>": <a list of tag values, for #e — a list of event ids, for #p — a list of pubkeys, etc.>
```

Author: As I said I made it like this because some people might be using more than one letter.

But you're right, it isn't worth the trouble. If that is needed we can add other methods later.

Reviewer: It does not make sense to create an invalid NIP just because some users might want to break NIP-01. If a future NIP adds support for filtering by multiple letters, we can implement a method for it then. That NIP should avoid using the '#' symbol, as it is reserved for single-letter tags.

### `nostr.req_add_tag_bin32(req: i32, tag_ptr: i32, value_ptr: i32) -> void`

Same as `req_add_tag` but the value is a 32-byte binary buffer that the host will convert to hex. `value_ptr` must point to a 32-byte buffer. This is useful for `#p`, `#e`, and other tag filters that need pubkey or event ID values.

Hosts may deduplicate filter values (authors, ids, kinds, and tag values) when the module adds the same value more than once via `req_add_*`.

### `nostr.req_set_limit(req: i32, limit: i32) -> void`

Sets the `"limit"` attribute of the filter.

### `nostr.req_set_since(req: i32, timestamp: i32) -> void`

Sets the `"since"` attribute of the filter.

### `nostr.req_set_until(req: i32, timestamp: i32) -> void`

Sets the `"until"` attribute of the filter.

> 

Reviewer: The `limit` and `timestamp` must be greater than zero. The runtime should validate that this condition is met.

If not, what should it do? ignore the call?

Author: Stop with an error? I don't know.

Reviewer: Sorry, but this hasn't been resolved. The runtime cannot add `-69` as a limit
to the filter. For example, in `rust-nostr/nostr`, the `limit` is an unsigned
number, so the runtime must ensure it's greater than 0.

What should the runtime do? I think adding a note that it should be greater than
0 is sufficient, the runtime should ignore the call if the value is not greater
than 0.

Reviewer: @fiatjaf

To cancel a live subscription call `nostr.drop(sub)`.

### `nostr.event_get_id(event_handle: i32) -> i32`

Returns a pointer to a 32-byte buffer containing the event ID.

### `nostr.event_get_id_hex(event_handle: i32) -> i32`

Returns a pointer to a 64-character string containing the hex event ID.

### `nostr.event_get_pubkey(event_handle: i32) -> i32`

Returns a pointer to a 32-byte buffer containing the public key of the event author.

### `nostr.event_get_pubkey_hex(event_handle: i32) -> i32`

Returns a pointer to a 64-character string containing the hex public key of the event author.

### `nostr.event_get_kind(event_handle: i32) -> i32`

Returns the kind integer directly.

### `nostr.event_get_created_at(event_handle: i32) -> i32`

Returns the unix timestamp directly.

### `nostr.event_get_content(event_handle: i32) -> i32`

Returns a pointer to a buffer containing the event content (follows the String Convention above).

### `nostr.event_get_tag_count(event_handle: i32) -> i32`

Returns the total number of tags on this event.

### `nostr.event_get_tag_item_count(event_handle: i32, tag_index: i32) -> i32`

> Returns the number of items in the tag at `tag_index`.

Reviewer: ```suggestion
Returns the number of items in the tag at `tag_index`. `0` if no tag in the given index
```

### `nostr.event_get_tag_item(event_handle: i32, tag_index: i32, item_index: i32) -> i32`

> Returns a pointer to a buffer containing the item at `(tag_index, item_index)` (follows the String Convention above).

Reviewer: ```suggestion
Returns a pointer to a buffer containing the item at `(tag_index, item_index)` (follows the String Convention above). `0` if no tag or value in the given index
```

### `nostr.event_get_tag_item_bin32(event_handle: i32, tag_index: i32, item_index: i32) -> i32`

> Same as `event_get_tag_item`, but returns a pointer to a 32-byte buffer of the item if it happened to be a pubkey or an event id; 0 otherwise.

Reviewer: ```suggestion
Same as `event_get_tag_item`, but returns a pointer to a 32-byte buffer of the item if it happened to be a pubkey or an event id; `0` if no tag or item in the given index or the value is not 64-byte hex string.
```

### `nostr.event_get_tag_item_by_name(event_handle: i32, name_ptr: i32, name_len: i32, item_index: i32) -> i32`

> Finds the first tag whose name (item 0) matches the string at `name_ptr/name_len`, then returns a pointer to a buffer containing item `item_index` from that tag; `0` if no matching tag is found.

Reviewer: you know what you should add here :)

Author: I don't know, I think it's better to just mention it once at the beginning.

Author: Nevermind, I'll add.

### `nostr.display(event: i32) -> void`

Render an event through the client's native note renderer. The event handle is not consumed.

### `nostr.log(ptr: i32, len: i32) -> void`

Emit a log message to the host's debug console or developer tooling. The string at `ptr/len` is the message.

### `nostr.drop(handle: i32) -> void`

Releases any handle: unconsumed request, active subscription (cancels it), individual event.

---

## Example programs and host code

- JavaScript host runtime example: https://viewsource.win/fiatjaf.com/nprogram/_/2e7d56d1536fcca8c49e7b2e99db8062f8df9ae1/~/src/nprogram-host.ts
- Rust example (basic, just reads from a relay): https://viewsource.win/fiatjaf.com/nprogram/_/2e7d56d1536fcca8c49e7b2e99db8062f8df9ae1/~/examples/basic/src/lib.rs
- Zig example (displays the history of all interactions between the current user and some other user): https://viewsource.win/fiatjaf.com/nprogram/_/2e7d56d1536fcca8c49e7b2e99db8062f8df9ae1/~/examples/interactions/main.zig
- TinyGo example (fetches all notifications, only display the unreplied ones): https://viewsource.win/fiatjaf.com/nprogram/_/2e7d56d1536fcca8c49e7b2e99db8062f8df9ae1/~/examples/unreplied/main.go

> - AssemblyScript example (fetches events from multiple relays and does a diff of the content): https://viewsource.win/fiatjaf.com/nprogram/_/2e7d56d1536fcca8c49e7b2e99db8062f8df9ae1/~/examples/old-trending/main.ts

Reviewer: Can you update the examples to point to the latest changes?

```suggestion
- JavaScript host runtime example: https://viewsource.win/fiatjaf.com/nprogram/_/master/~/src/nprogram-host.ts
- Rust example (basic, just reads from a relay): https://viewsource.win/fiatjaf.com/nprogram/_/master/~/examples/basic/src/lib.rs
- Zig example (displays the history of all interactions between the current user and some other user): https://viewsource.win/fiatjaf.com/nprogram/_/master/~/examples/interactions/main.zig
- TinyGo example (fetches all notifications, only display the unreplied ones): https://viewsource.win/fiatjaf.com/nprogram/_/master/~/examples/unreplied/main.go
- AssemblyScript example (fetches events from multiple relays and does a diff of the content): https://viewsource.win/fiatjaf.com/nprogram/_/master/~/examples/old-trending/main.ts
```

# NIP-5C

## Scrolls

`draft` `optional`

This NIP defines a standard for publishing self-contained WebAssembly programs as Nostr events ("scrolls"). Scrolls are composed of basic metadata (`name`, `description` and `icon`) and initial execution parameters defined as tags and a WASM binary, base64-encoded as the content. The binaries are executed in a sandbox inside a "host" (i.e. a proper Nostr client). Scrolls interact with Nostr only through a set of simple APIs provided by the host.

---

## Event Format

```yaml
{
  "kind": 1227,
  "content": "<base64-encoded WASM binary>",
  "tags":
    [
      ["name", "<name>"],
      ["description", "<description>"],
      ["icon", "<image-url>"],
      ["param", "<... (see below)>"]
    ]
}
```

## Favorite scrolls list

Clients can publish a list of favorite scrolls using a NIP-51 standard list event with kind `10027`. The list should include one `"e"` tag per scroll event, with optional relay hints and author hints.

## String Convention

> Unless noted otherwise (i.e. for hex pubkeys and ids), the module passes strings to the host as `(ptr: i32, len: i32)` pairs into its linear memory. The host returns strings (and other variable-length data) by returning an `i32` pointer to a buffer in linear memory whose first 4 bytes are the `u32_be` byte-length of the payload that follows.

Reviewer: ```suggestion
Unless noted otherwise (i.e. for hex pubkeys and ids), the module passes strings to the host as `(ptr: i32, len: i32)` pairs into its linear memory. The host returns strings (and other variable-length data) by returning an `i32` pointer to a buffer in linear memory whose first 4 bytes are the `u32` byte-length of the payload that follows.
```

> For example, if `nostr.get_content(handle)` returns `7165` that will be a pointer to a buffer at location `7165`. If that buffer starts with `[ 00 00 00 03 ]` that means it contains 3 bytes (aside from the initial four), so we can keep reading it `[ 00 00 00 03 66 6f 6f ]` and conclude that the string is `"foo"`.

Reviewer: This show the length provided as Little-Endian while above it say `u32_be`

Author: Oops.

Reviewer: you can resolve this, my suggestion follows below.

> 

Reviewer: If you would like, this is a clarification.

```suggestion

## Numbers Convention

All numbers written by the host must use **Little-endian** byte order, matching WASM's default endianness. This applies consistently: write string length prefixes as Little-endian numbers, write number parameters as Little-endian

```

The host must prefix every parameter with a single presence byte: `1` indicates the parameter is provided, and `0` indicates it is omitted. If a required parameter is prefixed with `0`, the program must panic.

| Type         | Encoding                                                              | Size (if provided) |
| ------------ | --------------------------------------------------------------------- | ------------------ |
| `public_key` | 32 bytes                                                              | 32 bytes           |
| `event`      | i32 handle                                                            | 4 bytes            |
| `string`     | u32_be length followed by UTF-8 bytes                                 | 4 + len bytes      |
| `number`     | i32                                                                   | 4 bytes            |
| `timestamp`  | Unix timestamp as u32_be                                              | 4 bytes            |

> | `relay`      | Relay URL, same encoding as string                                    | 4 + len bytes      |

Reviewer: Why is the string and timestamp is big-endian while the rest is little-endian?

WASM default is little-endian, I think it should always be LE.

---

> An unsigned integer of N bits, represented in N/8 bytes in little endian order. N is either 8, 16, or 32.
> https://www.wasm.com.cn/docs/binary-encoding/

> Little-endian byte ordering.
> https://webassembly.org/docs/portability/

> All values are read and written in [little endian](https://en.wikipedia.org/wiki/Endianness#Little-endian) byte order.
> https://www.w3.org/TR/wasm-core-2/#memory-instructions%E2%91%A0

Author: Yes, you are right. Please send a suggestion here.

Reviewer: you can resolve this, my suggestion follows below.

> | `relay`      | Relay URL, same encoding as string                                    | 4 + len bytes      |

Reviewer: ```suggestion
| Type         | Encoding                           | Size (if provided) |
| ------------ | -----------------------------------| ------------------ |
| `public_key` | 32 bytes                           | 32 bytes           |
| `event`      | i32 handle                         | 4 bytes            |
| `string`     | u32 length followed by UTF-8 bytes | 4 + len bytes      |
| `number`     | i32                                | 4 bytes            |
| `timestamp`  | Unix timestamp as u32              | 4 bytes            |
| `relay`      | Relay URL, same encoding as string | 4 + len bytes      |
```

Many of these expect and return handles. Handles are just `i32` numbers. The idea of dealing with handles and providing accessor functions is that the WASM programs don't have to include JSON parsers or any other form of encoder or decoder that would be necessary in order to send and receive structured data to and from the host. Instead they just deal with handles and request data to be extracted from these handles on their behalf to the host. This allows WASM programs to remain small and easy to write.

### `nostr.req_new() -> i32`

Create a new empty request handle.

### `nostr.req_add_author(req: i32, pubkey_ptr: i32) -> void`

Add a pubkey to the `authors` filter. It must be a pointer to a 32-byte buffer.

### `nostr.req_add_author_hex(req: i32, pubkey_hex_ptr: i32) -> void`

Same as `req_add_author` but with a hex-encoded pubkey string. Length is assumed to be 64 characters.

### `nostr.req_add_id(req: i32, id_ptr: i32) -> void`

Add an id to the `ids` filter. It must be a pointer to a 32-byte buffer.

### `nostr.req_add_id_hex(req: i32, id_hex_ptr: i32) -> void`

Same as `req_add_id` but with a hex-encoded id string. Length is assumed to be 64 characters.

> ### `nostr.req_add_kind(req: i32, kind: i32) -> void`

Reviewer: ```suggestion
### `nostr.req_add_kind(req: i32, kind: u32) -> void`
```

Add a kind integer to the `kinds` filter.

### `nostr.req_add_tag(req: i32, tag: i32, value_ptr: i32, value_len: i32) -> void`

Add a value to a tag filter. `tag` is the ASCII code of the letter (e.g. `112` is `"p"` and will be added to the filter as `"#p"`); `value_ptr/value_len` points to the tag value to match. The value is treated as a string.

### `nostr.req_add_tag_bin32(req: i32, tag: i32, value_ptr: i32) -> void`

Same as `req_add_tag` but the value is a pointer to the 32-byte binary buffer that the host will convert to hex. This is useful for `#p`, `#e`, and other tag filters that need pubkey or event ID values.

Hosts may deduplicate filter values (authors, ids, kinds, and tag values) when the module adds the same value more than once via `req_add_*`.

> ### `nostr.req_set_limit(req: i32, limit: i32) -> void`

Reviewer: ```suggestion
### `nostr.req_set_limit(req: i32, limit: u32) -> void`
```

Sets the `"limit"` attribute of the filter.

> ### `nostr.req_set_since(req: i32, timestamp: i32) -> void`

Reviewer: ```suggestion
### `nostr.req_set_since(req: i32, timestamp: u32) -> void`
```

Sets the `"since"` attribute of the filter.

> ### `nostr.req_set_until(req: i32, timestamp: i32) -> void`

Reviewer: ```suggestion
### `nostr.req_set_until(req: i32, timestamp: u32) -> void`
```

### `nostr.req_close_on_eose(req: i32) -> void`

Mark this request to automatically close after `EOSE`.

### `nostr.subscribe(req: i32) -> i32`

Sends the REQ to the target relays and returns a subscription handle. The request handle is consumed by this call — do not drop it separately.

The globally exported `on_event` function will be called with the results.

- The module must call `nostr.drop(event)` on each non-zero event handle when done with it.
- If `nostr.req_close_on_eose` was set, the host drops the subscription handle after the EOSE callback. The module must not drop it.

To cancel a live subscription call `nostr.drop(sub)`.

### `nostr.event_get_id(event_handle: i32) -> i32`

Returns a pointer to a 32-byte buffer containing the event ID.

### `nostr.event_get_id_hex(event_handle: i32) -> i32`

Returns a pointer to a 64-character string containing the hex event ID.

### `nostr.event_get_pubkey(event_handle: i32) -> i32`

Returns a pointer to a 32-byte buffer containing the public key of the event author.

### `nostr.event_get_pubkey_hex(event_handle: i32) -> i32`

Returns a pointer to a 64-character string containing the hex public key of the event author.

> ### `nostr.event_get_kind(event_handle: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_kind(event_handle: i32) -> u32`
```

Returns the kind integer directly.

> ### `nostr.event_get_created_at(event_handle: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_created_at(event_handle: i32) -> u32`
```

Returns the unix timestamp directly.

### `nostr.event_get_content(event_handle: i32) -> i32`

Returns a pointer to a buffer containing the event content (follows the String Convention above).

> ### `nostr.event_get_tag_count(event_handle: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_tag_count(event_handle: i32) -> u32`
```

Returns the total number of tags on this event.

> ### `nostr.event_get_tag_item_count(event_handle: i32, tag_index: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_tag_item_count(event_handle: i32, tag_index: i32) -> u32`
```

> ### `nostr.event_get_tag_item_count(event_handle: i32, tag_index: i32) -> u32`

Reviewer: ```suggestion
### `nostr.event_get_tag_item_count(event_handle: i32, tag_index: u32) -> u32`
```

Returns the number of items in the tag at `tag_index`; or `0` if such tag doesn't exist.

> ### `nostr.event_get_tag_item(event_handle: i32, tag_index: i32, item_index: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_tag_item(event_handle: i32, tag_index: u32, item_index: u32) -> i32`
```

Returns a pointer to a buffer containing the item at `(tag_index, item_index)` (follows the String Convention above); or `0` if such tag doesn't exist.

> ### `nostr.event_get_tag_item_bin32(event_handle: i32, tag_index: i32, item_index: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_tag_item_bin32(event_handle: i32, tag_index: u32, item_index: u32) -> i32`
```

Same as `event_get_tag_item`, but returns a pointer to a 32-byte buffer of the item if it happened to be a pubkey or an event id; `0` otherwise.

> ### `nostr.event_get_tag_item_by_name(event_handle: i32, name_ptr: i32, name_len: i32, item_index: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_tag_item_by_name(event_handle: i32, name_ptr: i32, name_len: i32, item_index: u32) -> i32`
```

Finds the first tag whose name (item 0) matches the string at `name_ptr/name_len`, then returns a pointer to a buffer containing item `item_index` from that tag; `0` if no matching tag is found.

> ### `nostr.event_get_tag_item_by_name_bin32(event_handle: i32, name_ptr: i32, name_len: i32, item_index: i32) -> i32`

Reviewer: ```suggestion
### `nostr.event_get_tag_item_by_name_bin32(event_handle: i32, name_ptr: i32, name_len: i32, item_index: u32) -> i32`
```
